internal/http.go

```diff
@@ -0,0 +1,206 @@
+	defer cancelShutdown()
+
+	if err := httpServer.Shutdown(gracefullCtx); err != nil {
+		reportLog.Info("http server shutdown error: %v\n", err)
+		defer os.Exit(1)
+	} else {
+		reportLog.Info("gracefully stopping http server\n")
+	}
+	defer os.Exit(0)
+}
+
+func downloadReport(w http.ResponseWriter, r *http.Request) {
+	if r.URL.Path != "/download" {
+		http.Error(w, "404 not found.", http.StatusNotFound)
+		return
+	}
+
+	if r.Method != "GET" {
+		http.Error(w, "method is not supported.", http.StatusMethodNotAllowed)
+		return
+	}
+
+	query := r.URL.Query()
+	start, present := query["start"]
+	if !present || len(start) == 0 {
+		fmt.Println("start time not present")
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	end, present := query["end"]
+	if !present || len(end) == 0 {
+		fmt.Println("end time not present")
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	conn, err := pgx.Connect(context.Background(), postgresURL)
+	if err != nil {
+		reportLog.Error(err, "database connection error:")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	dirName := "/tmp/curator-report"
+	error := os.MkdirAll(dirName, 0755)
+	if error != nil {
+		fmt.Println("error opening file:", err)
+		return
+	}
+
+	tarDirName := fmt.Sprintf("%s/%s-%s-koku-metrics.tar.gz",
+		dirName, strings.Join(start, ","), strings.Join(end, ","))
+
+	outputFile, err := os.OpenFile(tarDirName, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		log.Fatalln(err)
+	}
+	defer outputFile.Close()
+
+	writer, err := gzip.NewWriterLevel(outputFile, gzip.BestCompression)
+	if err != nil {
+		log.Fatalln(err)
+	}
+	defer writer.Close()
+
+	tw := tar.NewWriter(writer)
+	defer tw.Close()
+
+	for i := 0; i < 4; i++ {
+		sqlQuery := fmt.Sprintf("COPY (SELECT * FROM logs_%d WHERE interval_start >= '%s'::timestamp with time zone AND interval_end < '%s' "+
+			"::timestamp with time zone) TO STDOUT WITH CSV DELIMITER ',' HEADER", i, strings.Join(start, ","), strings.Join(end, ","))
+
+		outfile := fmt.Sprintf("/tmp/curator-report/%s-%s-koku-metrics.%d.csv", strings.Join(start, ","), strings.Join(end, ","), i)
+
+		ef, err := os.OpenFile(outfile, os.O_RDWR|os.O_CREATE, 0755)
+		if err != nil {
+			fmt.Println("error opening file:", err)
+			return
+		}
+
+		err = exporter(conn, ef, sqlQuery)
+		if err != nil {
+			fmt.Println("error exporting report:", err)
+		}
+		ef.Close()
+
+		err = addToArchive(tw, outfile)
+		if err != nil {
+			fmt.Println("error archiving file:", err)
+		}
+
+		err = os.Remove(outfile)
+		if err != nil {
+			fmt.Println("error removing file:", err)
+		}
+	}
+
+	defer conn.Close(context.Background())
+	w.WriteHeader(http.StatusOK)
+
+	_, err = w.Write([]byte("generated report"))
+	if err != nil {
+		log.Println("Failed writing HTTP response")
+	}
+}
+
+func exporter(conn *pgx.Conn, f *os.File, query string) error {
+	_, err := conn.PgConn().CopyTo(context.Background(), f, query)
```

Comment:
Let's avoid explicitly using context.Background outside of driver functions whenever we can and instead pass a context.Context parameter?